Perform additional checks in `PEHeaders`.

teo-tsirpanis opened the pull request:
Fixes #48419.
Supersedes #112830 for now.

See also https://github.com/dotnet/runtime/issues/112830#issuecomment-2677632672.

This PR updates the constructor of `PEHeaders` to perform some additional checks to detect files with an obviously invalid structure. Specifically:

* `CoffHeader.SizeOfOptionalHeader` is checked to be zero in object files.
* The size of the file is checked to fit as many as `CoffHeader.NumberOfSections` sections, in both object and image files.
* COFF files are rejected if `isLoadedImage` is set to `true`.

teo-tsirpanis committed:
Perform additional checks in `PEHeaders`.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeaders.cs:
```diff
@@ -301,7 +314,12 @@ private ImmutableArray<SectionHeader> ReadSectionHeaders(ref PEBinaryReader read
             {
-                builder.Add(new SectionHeader(ref reader));
+                var sectionHeader = new SectionHeader(ref reader);
+                if (isCoffOnly && sectionHeader.VirtualSize != 0)
```
https://github.com/dotnet/runtime/issues/112830#issuecomment-2677632672 said that `VirtualAddress` is the field that must be zero in object files, but [according to documentation](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#:~:text=This%20field%20is%20valid%20only%20for%20executable%20images%20and%20should%20be%20set%20to%20zero%20for%20object%20files.) it's `VirtualSize`.

steveharter replied:
That sounds right. For `VirtualAddress` the doc says "For object files, this field is the address of the first byte before relocation is applied; for simplicity, compilers should set this to zero.". So yeah, we probably don't want to enforce zero for object files. Even for executable files this may be zero for some sections like `.debug`.

teo-tsirpanis committed:
Fix typo.

teo-tsirpanis committed:
Update test.

teo-tsirpanis commented:
Most test failures were due to typos, but in one test we were opening an `.snk` file, which now (correctly) throws. That test was updated and moved to `PEHeaders`.

steveharter commented on src/libraries/System.Reflection.Metadata/tests/PortableExecutable/PEHeadersTests.cs:
```diff
@@ -50,6 +50,13 @@ public void Ctor_Streams()
         }
 
+        [Fact]
+        public void Ctor_InvalidFile()
```
What case is this catching from above, and related question is what coverage do we have on the new checks?

teo-tsirpanis replied:
It catches the "size of optional header is nonzero" case. I can add more tests.

MichalPetryka commented:
Should this be labeled as a breaking change?

MichalStrehovsky commented:
The list of heuristics I gave in https://github.com/dotnet/runtime/issues/112830#issuecomment-2677632672 is somewhat haphazard. Now that I'm looking at the actual code, I wonder if it would be better to do fewer of these validations (it also helps in having to write fewer tests since there's fewer codepaths that need to be covered).

The things that we _need_ to validate are things that would make us crash in bad ways or are _structurally_ undefined:

* `SizeOfOptionalHeader` - if this is non-zero, there is an optional header following the COFF header that we should presumably read or skip so that we don't read optional header data as section headers. But this is all undefined territories. So throw if it's non-zero.
* Checking `PointerToSymbolTable` and `Characteristics` is one of the haphazard heuristics I gave. We don't really care if these are nonsensical, so maybe we don't _need_ to look at them at all.
* Checking that we can read all the declared section headers and don't crash with a stream read exception is important, so we should do the validation.
* Checking the `VirtualSize` of the section doesn't really matter.

We'll allow more COFF object file-like files but looking at the APIs on `PEHeaders`, this should be fine:

* CoffHeader, SectionHeaders is fine - the values will be junk but the API doesn't judge. Trying to filter some of the junk wouldn't move the needle, even with the validation we'd let junk through.
* CoffHeaderStartOffset, IsCoffOnly, CorHeader, CorHeaderStartOffset, PEHeader, PEHeaderStartOffset all depend on this being an image so these APIs will be fine
* MetadataStartOffset and MetadataSize will be null because none of the junk SectionHeaders will have name `.cormeta`

The 99% use case of PEHeader is for getting MetadataStartOffset and MetadataSize so that we can get a metadata reader. And this is going to be pretty well defined if we do the above validation that prevents crashes only.

teo-tsirpanis committed:
Perform less checks.

teo-tsirpanis committed:
Update tests.

teo-tsirpanis commented:
Updated PR to check only for `SizeOfOptionalHeader` and `NumberOfSections` and added test coverage for these cases. I'm not sure I agree with making less checks; we don't have an explicit indicator that a file is not a COFF file, so we should check for as many implicit indicators as reasonably possible in order to fail with a clear message, but for the invalid files that pass these two checks users will either get an exception or a negative `HasMetadata` result later either way.

teo-tsirpanis commented on src/libraries/System.Reflection.Metadata/src/System/Reflection/PortableExecutable/PEHeaders.cs:
```diff
@@ -93,13 +93,21 @@ public PEHeaders(Stream peStream, int size, bool isLoadedImage)
 
-            if (!isCoffOnly)
+            if (isCoffOnly)
+            {
```
```suggestion
            {
                if (isLoadedImage)
                {
                    throw new BadImageFormatException(SR.InvalidPESignature);
                }
```

What do you think about adding this check (a bit earlier)? The OS can't load an object file, can it?

MichalStrehovsky replied:
Adding such check sounds good to me.

teo-tsirpanis committed:
Throw when reading loaded images without a PE header.

teo-tsirpanis commented:
> Should this be labeled as a breaking change?

I don't think it's necessary. Reading an invalid PE file might already throw in certain cases (for example if it has a DOS stub but not a PE signature afterwards, or if it causes out of bounds reads), and this PR broadens these cases.
